### PR TITLE
QuickContact in GroupMembersDialog

### LINKED ISSUE #2837: Contact numbers copyable
Someone created a group conversation. You don't have every contact saved in your address book. You want to copy the number of one contact and want to create a new contract entry.

Problem: I didn't find any chance to copy a number of a contact.

### PR DESCRIPTION
**Pull request by chr1shaefn3r**
Added onClickListener to the GroupMembersDialog which gives the same functionality as taps on avatars.

Fixes #2837

### CODE REVIEW

**Comment by mcginty** [comment]
At least in Gingerbread, clicking on a group contact results in a NPE so marking this as higher priority as it solves a crash bug.

File: src/org/thoughtcrime/securesms/GroupMembersDialog.java [hunk 1]
```diff
@@ -56,20 +52,12 @@ public void onPostExecute(Recipients members) {
...
-    List<String> recipientStrings = new LinkedList<>();
-    recipientStrings.add(context.getString(R.string.GroupMembersDialog_me));
-
-    for (Recipient recipient : members.getRecipientsList()) {
-      if (!isLocalNumber(recipient)) {
-        recipientStrings.add(recipient.toShortString());
-      }
-    }
-
+    GroupMembers groupMembers = new GroupMembers(context, members);
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/GroupMembersDialog.java]: i think it's fine the way it is, we don't need a whole new class that is another way to represent recipients

File: src/org/thoughtcrime/securesms/GroupMembersDialog.java [hunk 2]
```diff
@@ -56,20 +52,12 @@ public void onPostExecute(Recipients members) {
...
     AlertDialogWrapper.Builder builder = new AlertDialogWrapper.Builder(context);
     builder.setTitle(R.string.ConversationActivity_group_members);
     builder.setIconAttribute(R.attr.group_members_dialog_icon);
     builder.setCancelable(true);
-    builder.setItems(recipientStrings.toArray(new String[]{}), null);
+    builder.setItems(groupMembers.getRecipientStrings(), new GroupMembersOnClickListener(context, groupMembers));
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/GroupMembersDialog.java]: make the listener an inner class

File: src/org/thoughtcrime/securesms/recipients/Recipient.java [hunk 3]
```diff
@@ -89,6 +91,10 @@ public void onFailure(Throwable error) {
     this.contactPhoto               = contactPhoto;
   }
 
+  public synchronized boolean isAPhonebookContact() {
+    return (this.contactUri != null);
+  }
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/recipients/Recipient.java]: take this out

File: src/org/thoughtcrime/securesms/recipients/Recipient.java [hunk 4]
```diff
@@ -154,6 +160,13 @@ public synchronized Bitmap getGeneratedAvatar(Context context) {
     return this.generatedAvatar;
   }
 
+  public Intent getInsertOrEditIntent() {
+    final Intent intent = new Intent(Intent.ACTION_INSERT_OR_EDIT);
+    intent.putExtra(ContactsContract.Intents.Insert.PHONE, getNumber());
+    intent.setType(ContactsContract.Contacts.CONTENT_ITEM_TYPE);
+    return intent;
+  }
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/recipients/Recipient.java]: this is a model, it shouldn't know anything about intents

File: src/org/thoughtcrime/securesms/GroupMembersDialog.java (hunk 2 above)
Comment (chr1shaefn3r) [inline src/org/thoughtcrime/securesms/GroupMembersDialog.java]: - done -

File: src/org/thoughtcrime/securesms/recipients/Recipient.java (hunk 3 above)
Comment (chr1shaefn3r) [inline src/org/thoughtcrime/securesms/recipients/Recipient.java]: done

File: src/org/thoughtcrime/securesms/recipients/Recipient.java (hunk 4 above)
Comment (chr1shaefn3r) [inline src/org/thoughtcrime/securesms/recipients/Recipient.java]: agreed, moved to RecipientUtil to keep the creation of this intent in one place.

File: src/org/thoughtcrime/securesms/GroupMembersDialog.java (hunk 1 above)
Comment (chr1shaefn3r) [inline src/org/thoughtcrime/securesms/GroupMembersDialog.java]: The problem I needed to solve was to match the index I get from the Dialog via the onClick callback to the correct recipient. Second priority was to still be able to have 'Me' on top of the list. Therefore I created this new class fulfil those two requirements.

File: src/org/thoughtcrime/securesms/groups/GroupMembers.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * Wraps a List of Recipient (just like @class Recipients),
+ * but with focus on the order of the Recipients.
+ * So that the order of the RecipientStrings[] matches
+ * the internal order.
+ *
+ * @author: Christoph Haefner
+ */
+public class GroupMembers {
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/groups/GroupMembers.java]: make it an inner class

File: src/org/thoughtcrime/securesms/groups/GroupMembers.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  private static final String TAG = GroupMembers.class.getSimpleName();
+
+  private LinkedList<Recipient> sortedMembers;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/groups/GroupMembers.java]: final and initialize

File: src/org/thoughtcrime/securesms/groups/GroupMembers.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    sortedMembers = new LinkedList<>();
+
+    List<Recipient> recipientList = recipients.getRecipientsList();
+    for (Recipient recipient : recipientList) {
+      if (isLocalNumber(recipient)) {
+        sortedMembers.push(recipient);
+      } else {
+        sortedMembers.add(recipient);
+      }
+    }
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/groups/GroupMembers.java]: put this in the constructor

File: src/org/thoughtcrime/securesms/groups/GroupMembers.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  private List<String> createListOfRecipientStrings() {
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/groups/GroupMembers.java]: doesn't look like you need a separate private method for this

File: src/org/thoughtcrime/securesms/groups/GroupMembers.java (hunk 5 above)
Comment (chr1shaefn3r) [inline src/org/thoughtcrime/securesms/groups/GroupMembers.java]: done

File: src/org/thoughtcrime/securesms/groups/GroupMembers.java (hunk 6 above)
Comment (chr1shaefn3r) [inline src/org/thoughtcrime/securesms/groups/GroupMembers.java]: done

File: src/org/thoughtcrime/securesms/groups/GroupMembers.java (hunk 7 above)
Comment (chr1shaefn3r) [inline src/org/thoughtcrime/securesms/groups/GroupMembers.java]: done

File: src/org/thoughtcrime/securesms/groups/GroupMembers.java (hunk 8 above)
Comment (chr1shaefn3r) [inline src/org/thoughtcrime/securesms/groups/GroupMembers.java]: done

File: src/org/thoughtcrime/securesms/GroupMembersDialog.java [hunk 9]
```diff
@@ -79,15 +73,83 @@ public void display() {
...
+   * but with focus on the order of the Recipients.
+   * So that the order of the RecipientStrings[] matches
+   * the internal order.
+   *
+   * @author: Christoph Haefner
+   */
+  private class GroupMembers {
+    private final String TAG = GroupMembers.class.getSimpleName();
+
+    private final LinkedList<Recipient> sortedMembers;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/GroupMembersDialog.java]: members = new LinkedList<>();

File: src/org/thoughtcrime/securesms/GroupMembersDialog.java [hunk 10]
```diff
@@ -79,15 +73,83 @@ public void display() {
...
+
+    public GroupMembers(Recipients recipients) {
+      sortedMembers = new LinkedList<>();
+
+      List<Recipient> recipientList = recipients.getRecipientsList();
+      for (Recipient recipient : recipientList) {
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/GroupMembersDialog.java]: for (Recipient recipient : recipients.getRecipientsList())

File: src/org/thoughtcrime/securesms/GroupMembersDialog.java [hunk 11]
```diff
@@ -79,15 +73,83 @@ public void display() {
...
+
+    public String[] getRecipientStrings() {
+      return createListOfRecipientStrings().toArray(new String[sortedMembers.size()]);
+    }
+
+    public Recipient get(int index) {
+      return sortedMembers.get(index);
+    }
+
+    private List<String> createListOfRecipientStrings() {
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/GroupMembersDialog.java]: you don't need a private method for this

File: src/org/thoughtcrime/securesms/util/RecipientUtil.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+import android.content.Intent;
+import android.provider.ContactsContract;
+
+import org.thoughtcrime.securesms.recipients.Recipient;
+
+/**
+ * @author Christoph Häfner
+ */
+public class RecipientUtil {
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/util/RecipientUtil.java]: we don't need a whole new class for this

File: src/org/thoughtcrime/securesms/GroupMembersDialog.java (hunk 9 above)
Comment (chr1shaefn3r) [inline src/org/thoughtcrime/securesms/GroupMembersDialog.java]: done

File: src/org/thoughtcrime/securesms/GroupMembersDialog.java (hunk 10 above)
Comment (chr1shaefn3r) [inline src/org/thoughtcrime/securesms/GroupMembersDialog.java]: done

File: src/org/thoughtcrime/securesms/GroupMembersDialog.java (hunk 11 above)
Comment (chr1shaefn3r) [inline src/org/thoughtcrime/securesms/GroupMembersDialog.java]: done

File: src/org/thoughtcrime/securesms/util/RecipientUtil.java (hunk 12 above)
Comment (chr1shaefn3r) [inline src/org/thoughtcrime/securesms/util/RecipientUtil.java]: I'm out of idea where to put this, therefore I removed it.

**Comment by rhodey** [comment]
clicking on a group contact still results in an NPE, Nexus 5 Lollipop 5.1. higher priority still for sure.

```
05-07 10:45:56.609  23925-23925/org.thoughtcrime.securesms.sushi D/AndroidRuntime﹕ Shutting down VM
05-07 10:45:56.609  23925-23925/org.thoughtcrime.securesms.sushi E/AndroidRuntime﹕ FATAL EXCEPTION: main
    Process: org.thoughtcrime.securesms.sushi, PID: 23925
    java.lang.NullPointerException: Attempt to invoke interface method 'void com.afollestad.materialdialogs.MaterialDialog$ListCallback.onSelection(com.afollestad.materialdialogs.MaterialDialog, android.view.View, int, java.lang.CharSequence)' on a null object reference
            at com.afollestad.materialdialogs.MaterialDialog.onItemClick(MaterialDialog.java:169)
            at android.widget.AdapterView.performItemClick(AdapterView.java:305)
            at android.widget.AbsListView.performItemClick(AbsListView.java:1146)
            at android.widget.AbsListView$PerformClick.run(AbsListView.java:3053)
            at android.widget.AbsListView$3.run(AbsListView.java:3860)
            at android.os.Handler.handleCallback(Handler.java:739)
            at android.os.Handler.dispatchMessage(Handler.java:95)
            at android.os.Looper.loop(Looper.java:135)
            at android.app.ActivityThread.main(ActivityThread.java:5254)
            at java.lang.reflect.Method.invoke(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:372)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698)
05-07 10:45:57.770  23925-23985/org.thoughtcrime.securesms.sushi W/WebSocketConnection﹕ Sending keep alive...
```

**Pull request closed without merging**